tests/ui/match/ref_pat_eat_one_layer_2024/ref_pat_eat_one_layer_2024.rs

```diff
@@ -0,0 +1,32 @@
+//@ run-pass
+//@ edition: 2024
+//@ compile-flags: -Zunstable-options
+#![allow(incomplete_features)]
+#![feature(ref_pat_eat_one_layer_2024)]
+
+pub fn main() {
+    if let Some(Some(&x)) = &Some(&Some(0)) {
+        let _: u32 = x;
+    }
+    if let Some(Some(&x)) = &Some(Some(&0)) {
+        let _: &u32 = x;
+    }
+    if let Some(Some(&&x)) = &Some(Some(&0)) {
+        let _: u32 = x;
+    }
+    if let Some(&Some(x)) = &Some(Some(0)) {
+        let _: u32 = x;
+    }
+    if let Some(Some(&mut x)) = &mut Some(&mut Some(0)) {
+        let _: u32 = x;
+    }
+    if let Some(Some(&x)) = &Some(&Some(0)) {
+        let _: u32 = x;
+    }
+    if let Some(&mut Some(&x)) = &Some(&mut Some(0)) {
+        let _: u32 = x;
+    }
+    if let Some(&Some(&mut x)) = &mut Some(& Some(0)) {
+        let _: u32 = x;
+    }
```

Comment:
These two are weird, shouldn't we eat references in the correct order?

Reply:
Does this typecheck?
```rust
if let Some(&Some(x)) = &mut Some(&Some(0)) {
    let _: &mut u32 = x;
}
```

Reply:
> Does this typecheck?

Not any longer, pushed a fix so we now properly downgrade the mutability of the inherited reference

> These two are weird, shouldn't we be eating references in the same order as in the type?

No, that would be unnecessarily restrictive, as demonstrated by this newly added test:

```rust
if let Some(Some(&mut x)) = &Some(Some(&mut 0)) {
    let _: &u32 = x;
}
```

(`let &mut x = &&mut 0` does not work currently, but I think it should, will try to fix)

Reply:
Hm, I would not expect these tests to work personally.

I would expect something like: `let &mut x = <expr>; x: T` iff `let x = <expr>; x: &mut T`, no?

Reply:
I don't see any good reason we should restrict it to that, when the currently implemented rule (if the pattern can match against the inherited reference, do so; otherwise match against a reference in the scrutinee type) is strictly more permissive.

Reply:
Wouldn't `let Some(&x) = &Some(&mut 0)` work just as well?

Reply:
@Nadrieril no, that would make `x` an `&mut {integer}` instead of an `&{integer}`, and then the compiler yells that you can't move out of a shared reference

Reply:
Oh I see. Well, I'm happy to merge your design, this is an implementation review not a design review. I'll just need some time to wrap my head around the new commits